Make empty c10-full

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46092 Make empty c10-full**

Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Update: with @swolchok's fixes to c10::optional, the regression went down to 2%. We decided to not land #46598.

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

smessmer committed:
Make empty c10-full

Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/MetaTensor.cpp:
```diff
@@ -7,34 +7,29 @@ namespace native {
+  c10::optional<Layout> layout,
+  c10::optional<Device> device,
+  c10::optional<bool> pin_memory,
+  c10::optional<c10::MemoryFormat> memory_format
```
It's cool that we managed to kill the duping, though we had better have handled it somewhere else now :)

smessmer replied:
you mean the assertion? Hm I guess that would be the task of the C++ frontend now since that's where we unpack the TensorOptions. I don't think it's currently doing it though.

smessmer replied:
ah good point, the C++ frontend actually does do it already.

ezyang commented on aten/src/ATen/native/MetaTensor.cpp:
```diff
@@ -7,34 +7,29 @@ namespace native {
-    DispatchKeySet{DispatchKey::Meta, options.computeDispatchKey()},
-    dtype,
+    DispatchKeySet{DispatchKey::Meta, computeDispatchKey(dtype, layout, device)},
+    dtype.has_value() ? scalarTypeToTypeMeta(*dtype) : get_default_dtype(),
```
One thing that is annoying about the change here is it is now possible for `computeDispatchKey` and `dtype.has_value() ? scalarTypeToTypeMeta(*dtype) : get_default_dtype()` to diverg in their behavior (since computeDispatchKey is going to internally recompute the defaulting itself. It would be better if optional dtypes are resolved once up top, then `computeDispatchKey` can make use of the resolved values directly.

smessmer replied:
I switched it to use `dtype_or_default` which is the same defaulting logic used by `computeDispatchKey`.

ezyang commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -368,9 +364,18 @@ Tensor empty_like(
+    dtype_opt.value_or_else([&] {return typeMetaToScalarType(self.options().dtype());}),
+    layout_opt.value_or(self.options().layout()),
+    device_opt.value_or(self.options().device()),
+    pin_memory_opt.value_or(self.options().pinned_memory()),
```
This is misleading because there is never actually pinned memory set on `self.options`

```
TensorOptions Tensor::options() const {
  return TensorOptions().dtype(dtype())
                        .device(device())
                        .layout(layout());
}
```

ezyang commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -368,9 +364,18 @@ Tensor empty_like(
+  return at::empty(size,
+    dtype_opt.value_or_else([&] {return typeMetaToScalarType(self.options().dtype());}),
+    layout_opt.value_or(self.options().layout()),
+    device_opt.value_or(self.options().device()),
```
NB: this is only semantics preserving because `self.options()` is guaranteed to have settings for dtype, layout and device. If it were possible for these to be nullopt then this transformation would not be semantics preserving

smessmer replied:
good point. Do you think I can keep relying on it or should I rewrite this to use `self.options().layout_opt()`?

smessmer replied:
nvm, I rewrote it

ezyang commented on aten/src/ATen/native/mkldnn/TensorFactories.cpp:
```diff
@@ -4,23 +4,23 @@ namespace at { namespace native {
      "'memory_format' argument is incompatible with mkldnn tensor");
+  TORCH_CHECK(
+    !layout.has_value(),
+    "'layout' argument is incompatible with mkldnn tensor");
```
New!

smessmer replied:
yes but I thought adding it makes sense. Do you think we should not add it?

smessmer replied:
ok I removed it

ezyang commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -116,9 +119,9 @@ SparseTensor new_with_dims_and_tensor_sparse(
-  TORCH_CHECK(!options.pinned_memory(), "Only dense CPU tensors can be pinned");
-  return new_with_dims_sparse(size.size(), 0, size, options);
+Tensor empty_sparse(IntArrayRef size, c10::optional<ScalarType> dtype, c10::optional<Layout> layout, c10::optional<Device> device, c10::optional<bool> pin_memory, c10::optional<MemoryFormat> optional_memory_format) {
+  TORCH_CHECK(!pin_memory.has_value() || !*pin_memory, "Only dense CPU tensors can be pinned");
```
I saw this idiom in a few places and it was weird to me that you didn't just use the `blah_or_default` functions here. I guess it's something to do with asserts?

smessmer replied:
I found this more expressive since it's a precondition check for the function and this way it explicitly states the allowed values for the argument. With `pinned_memory_or_default` you'd have to look up that function to know if the `nullopt` case is allowed or not.

ezyang commented on c10/core/TensorOptions.h:
```diff
@@ -21,19 +21,19 @@ namespace c10 {
 
 inline Device device_or_default(c10::optional<Device> device) {
-  return device.has_value() ? *device : Device(kCPU);
+  return device.value_or_else([] {return Device(kCPU);});
```
`value_or_else` here probably isn't buying you much

smessmer replied:
you're probably right, the constructor is relatively cheap. But I wanted to be on the safe side and it definitely doesn't cost anything.

ezyang commented on c10/util/Optional.h:
```diff
@@ -508,6 +510,22 @@ class optional : private OptionalBase<T> {
   }
 
+  template <class F>
+  constexpr T value_or_else(F&& func) const& {
```
Isn't this, like, super cutting edge? I didn't see `value_or_else` in C++20, looks like there's some WG paper http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2218r0.pdf submitted literally like a month ago. Don't we have an implicit goal that c10::optional should track the standard std::optional?

smessmer replied:
yes it's relatively new. We can remove it again if the proposal doesn't make it in or if we want to replace c10::optional with std::optional before that. Should be a relatively simple string replacement to remove it if we need to.

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

bhosmer reviewed (approved):
[edit: at least some of the test failures look real, happy to re-review if they need nontrivial further changes ofc]

Nice to have empty() back in the c10-full family 😁 one nontrivial suggestion inline (I'd pitch removing that new assert), otherwise just formatting notes.

Oh also I can't remember, were you able to confirm that this is ~instruction-count neutral?

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 771.5k to 802.4k, which is a 4% regression.
Bench before this PR: P145725335
Bench after this PR: P145727341


Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 771.5k to 802.4k, which is a 4% regression.
Bench before this PR: P145725335
Bench after this PR: P145727341


Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Update: with @swolchok's fixes to c10::optional, the regression went down to 2%. We decided to not land #46598.

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]

smessmer committed:
Update on "Make empty c10-full"


Make empty c10-full without using hacky-wrapper, i.e. port the kernel to the new style signature.

This PR also changes the signature of some helpers called by empty to the new style.

Bench summary: With noisy symbols removed, instruction count goes from 661k to 692k, which is a 4-5% regression.
This regression is fixed in #46598 and a combination of both PRs actually improves perf by 2-3%, see P146297481

Update: with @swolchok's fixes to c10::optional, the regression went down to 2%. We decided to not land #46598.

Differential Revision: [D24219944](https://our.internmc.facebook.com/intern/diff/D24219944/)

[ghstack-poisoned]
